[scroll-animations] Implement basic view timeline progress calculation

nmoucht opened the pull request:
#### 1e0887d7f58e38cfb3f720ecb18dfca4ba43cc9f
<pre>
[scroll-animations] Implement basic view timeline progress calculation
<a href="https://bugs.webkit.org/show_bug.cgi?id=279799">https://bugs.webkit.org/show_bug.cgi?id=279799</a>
<a href="https://rdar.apple.com/136115845">rdar://136115845</a>

Reviewed by Simon Fraser.

Implement basic view timeline progress calculation based on the spec:
<a href="https://drafts.csswg.org/scroll-animations-1/#view-timeline-progress.">https://drafts.csswg.org/scroll-animations-1/#view-timeline-progress.</a> Also add a way
to find the source element for the ViewTimeline, based on the spec:
<a href="https://drafts.csswg.org/scroll-animations-1/#dom-scrolltimeline-source.">https://drafts.csswg.org/scroll-animations-1/#dom-scrolltimeline-source.</a>

* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::source const):
(WebCore::ViewTimeline::sourceRenderer const):
(WebCore::ViewTimeline::computeViewTimelineData const):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: <a href="https://commits.webkit.org/284126@main">https://commits.webkit.org/284126@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/f3f5f568a55f9ddeeed29cd2a49a94688dac5702

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/68506 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/47898 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/21165 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/72575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/19651 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/70623 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/55694 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/19467 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/5/builds/72575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/13082 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/71573 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/43792 "Passed tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/59158 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/5/builds/72575 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/40459 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/16579 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/18008 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/62416 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/16926 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/74269 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/12477 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/16188 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/62131 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/12516 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/59236 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/62156 "Passed tests") | 
| | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/10085 "Passed tests") | [  ~~🧪 mac-intel-wk2~~](https://ews-build.webkit.org/#/builders/119/builds/3707 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/10435 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/43699 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/44773 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/45967 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/44515 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -138,4 +142,75 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+
+    Ref document = m_subject->document();
+
+    CheckedPtr subjectRenderer = m_subject->renderer();
```
This makes me think we need some WPT that try to use scroll timeline on something with `display:contents` (which means the element won't have a renderer, but its children will).

nmoucht replied:
Added this to the list in the quip.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -138,4 +142,75 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+
+    // https://drafts.csswg.org/scroll-animations-1/#dom-scrolltimeline-source
+    // Determine source renderer by looking for the nearest ancestor that establishes a scroll container
+    return subjectRenderer->enclosingScrollableContainerForSnapping();
```
We should rename "for snapping"; it's confusing to see scroll snap stuff here. Please do that in this PR (or a prequel rename PR).

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -138,4 +142,75 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+
+    CheckedPtr<RenderBox> sourceRenderBox = sourceRenderer();
+
+    if (!sourceRenderBox)
```
Repetitive with the function above.

nmoucht replied:
These checks are still necessary in sourceRenderer() since it is used by source() as well.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -138,4 +142,75 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+    float currentScrollOffset = 0;
+    float scrollContainerSize;
+    if (sourceRenderBox->element() == document->documentElement()) {
+        auto& frameView = sourceRenderBox->view().frameView();
```
FrameView is a ScrollableArea so we should be able to write this code in terms of ScrollableArea, instead of branching.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -138,4 +142,75 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+        scrollContainerSize = axis() == ScrollAxis::Block ? sourceRenderBox->borderBoxRect().height() : sourceRenderBox->borderBoxRect().width();
+    }
+
+    float subjectOffset = axis() == ScrollAxis::Block ? subjectRenderBox->offsetTop() : subjectRenderBox->offsetLeft();
```
I don't think `offsetTop` is right here; that's relative to some ancestor, but not necessarily the scroller. Make a test case with a `position:relative` inside a `position:relative` inside a scroller.

I think scroll anchoring has the logic you need here.

nmoucht replied:
Changed to use `ScrollAnchoringController::computeOffsetFromScrollableArea`.

smfr commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -47,6 +47,12 @@ struct ViewTimelineInsets {
 
 class ViewTimeline final : public ScrollTimeline {
+    struct ViewTimelineData {
+        float scrollContainerSize;
```
Initialize these to zero.

graouts replied:
Could this struct be private? You could also rename it to just `Data` since it's within the `ViewTimeline` scope.

nmoucht replied:
Fixed.

graouts commented on Source/WebCore/animation/ViewTimeline.h:
```diff
@@ -59,7 +59,18 @@ class ViewTimeline final : public ScrollTimeline {
+        float scrollContainerSize = 0;
+        float subjectOffset = 0;
+        float currentScrollOffset = 0;
+        float coverRangeEnd = 0;
```
I think that the `{ 0 }` style is preferred, at least that's where the recent webkit-dev conversation is pointing to.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -140,4 +142,67 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+    if (!m_subject)
+        return nullptr;
+
+    Ref document = m_subject->document();
```
`document` is unused.

nmoucht replied:
Removed.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -140,4 +142,67 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
 
+Element* ViewTimeline::source() const
+{
+    if (auto* sourceRender = sourceRenderer())
```
I would suggest flipping this around; keep a reference to the element, and derive the renderer from it. Renderers can change (`display:none` toggled etc).

nmoucht replied:
We aren't holding a reference here just getting the result from `enclosingScrollableContainer`.

smfr commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -140,4 +142,67 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+    float currentScrollOffset = axis() == ScrollAxis::Block ? sourceScrollableArea->scrollPosition().y() : sourceScrollableArea->scrollPosition().x();
+    float scrollContainerSize = axis() == ScrollAxis::Block ? sourceScrollableArea->visibleHeight() : sourceScrollableArea->visibleWidth();
+
+    auto offsetFromScroller = ScrollAnchoringController::computeOffsetFromScrollableArea(*subjectRenderBox, *sourceScrollableArea, false);
```
Better to use an enum than pass a `false` value here. Hard to tell what it means.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollAnchoringController.cpp:
```diff
@@ -92,18 +92,18 @@ void ScrollAnchoringController::invalidateAnchorElement()
         return renderLayerScrollableArea->layer().renderer().absoluteBoundingBoxRect();
 
-    return IntRect(downcast<LocalFrameView>(downcast<ScrollView>(scrollableArea)).layoutViewportRect());
+    return shouldIncludeLayoutViewportRect ? IntRect(downcast<LocalFrameView>(downcast<ScrollView>(scrollableArea)).layoutViewportRect()) : IntRect();
```
Why does it return an empty rect instead of the FrameView's bounds or something? I'm also not sure what it means to "include" the layout viewport.

Seems like it would be cleaner to just return the FrameView's bounds, and have the caller deal with the behavior divergence.

nmoucht replied:
Renamed this to shouldIncludeFrameViewLocation. Made this only used in `ScrollAnchoringController::computeOffsetFromScrollableArea`.

nmoucht committed:
[scroll-animations] Implement basic view timeline progress calculation
https://bugs.webkit.org/show_bug.cgi?id=279799
rdar://136115845

Reviewed by Simon Fraser.

Implement basic view timeline progress calculation based on the spec:
https://drafts.csswg.org/scroll-animations-1/#view-timeline-progress. Also add a way
to find the source element for the ViewTimeline, based on the spec:
https://drafts.csswg.org/scroll-animations-1/#dom-scrolltimeline-source.

* Source/WebCore/animation/ViewTimeline.cpp:
(WebCore::ViewTimeline::source const):
(WebCore::ViewTimeline::sourceRenderer const):
(WebCore::ViewTimeline::computeViewTimelineData const):
* Source/WebCore/animation/ViewTimeline.h:

Canonical link: https://commits.webkit.org/284126@main

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -140,4 +142,65 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
 
+Element* ViewTimeline::source() const
+{
+    if (auto* sourceRender = sourceRenderer())
```
This should have been `CheckedPtr` I believe.

graouts commented on Source/WebCore/animation/ViewTimeline.cpp:
```diff
@@ -140,4 +142,65 @@ AnimationTimeline::ShouldUpdateAnimationsAndSendEvents ViewTimeline::documentWil
+    if (!subjectRenderBox)
+        return { };
+
+    auto* sourceScrollableArea = scrollableAreaForSourceRenderer(sourceRenderer(), document);
```
`CheckedPtr`
